feat: improve releases page

BinaryMuse opened the pull request:
This PR updates the releases page:

* Separates regular and nightly releases into separate pages

    Now that we are starting to generate nightly releases regularly, the releases page quickly gets filled with releases that a majority of Electron users likely don't care about. Two new links at the top of the releases page allow the user to select either regular releases (the default) or nightly releases.

* Paginates the releases list

    The releases list for regular releases was growing quite long, slowing down page rendering and increasing the time it took for JavaScript to kick in and do things like change the date/times into relative times (e.g. "4 days ago"). Also, since releases can contain a long list of release notes, the page was hard to scan. The paginated version of the page defaults to 5 releases per page.

I don't have a lot of strongly-held opinions on layout/style, so I'm very open to suggestions.

/cc @HashimotoYT 

- [x] Implement separate pages for regular & nightly releases
- [x] Implement pagination for releases
- [x] Update releases page tests

Fixes https://github.com/electron/electronjs.org/issues/1214
Fixes https://github.com/electron/electronjs.org/issues/1071

---

![image](https://user-images.githubusercontent.com/189606/51064162-296f4c80-15b3-11e9-922a-00209107bbf3.png)

---

![image](https://user-images.githubusercontent.com/189606/51064172-30965a80-15b3-11e9-9a52-0b6bdf5fe810.png)

---

![image](https://user-images.githubusercontent.com/189606/51064184-40ae3a00-15b3-11e9-8521-e960439dd707.png)

HashimotoYT commented:
## Option A

Sidebar with choosing release channel.

![first prototype](https://user-images.githubusercontent.com/24681191/51073681-6acc2e80-167d-11e9-92d7-3de93f619ef1.png)

---

## Option B

Same as Option A but with description for release channel.

![second prototype](https://user-images.githubusercontent.com/24681191/51073683-7ae40e00-167d-11e9-8fcb-6b2498f7cba1.png)

---

Identically for Option A and B.

## Sticky bar and multiusability

On sidebar we can add buttons like `Back to index` or `Subscribe RSS` or something similar to this.

![more buttons](https://user-images.githubusercontent.com/24681191/51073697-9ea75400-167d-11e9-8aba-c5ff2580bd23.png)

Why make sidebar sticky? 🤷‍♀️

## Active Element

For showing element who active we can use boring gray line what changes to blue `#4078C0` if it active.

| Unactive | Active |
| ------ | ------ |
| ![unactive release tag](https://user-images.githubusercontent.com/24681191/51073689-9222fb80-167d-11e9-93ac-02eca6cb1a5d.png) | ![active release tag](https://user-images.githubusercontent.com/24681191/51073692-9818dc80-167d-11e9-99c1-59f504c875a9.png)

BinaryMuse commented:
Alright, I've done the following:

* Split the releases into stable, beta, and nightly
* Fixed the previous/next button when at the first and last page
* Changed the URLs to `/releases/stable`, `/releases/beta`, and `/releases/nightly`, and updated links throughout the site. `/releases` now 301's to `/releases/stable`
* Removed nightly releases from the version history

BinaryMuse commented:
> Title not contain chosen release channel. It's think it should looks like it: Stable Releases | Electron.

Agreed. I localized the title in this case do you think we should do this?

> Releases text on navbar not contain active element

Fixed.

HashimotoYT commented on routes/releases/index.js:
```diff
@@ -1,3 +1,72 @@
+    const localizedReleasesType = req.context.localized.releases[localizedReleasesKey]
+
+    res.render('releases', Object.assign({}, req.context, {
+      page: { title: `${localizedReleasesType} | Electron Blog` },
```
🔥 Blog

```suggestion
      page: { title: `${localizedReleasesType} | Electron` },
```

BinaryMuse commented:
> Yes, we localize title on other pages

I think for a lot of the routes that specify `page.title`, the title isn't localized:

```
$ ack -C1 'title:' routes
routes/releases/index.js
67-    res.render('releases', Object.assign({}, req.context, {
68:      page: { title: `${localizedReleasesType} | Electron Blog` },
69-      releasesPage: new ReleasesPage(type, selectedReleases, req.query)

routes/userland/show.js
35-    page: {
36:      title: `${report.title} | Electron`
37-    }

routes/docs/show.js
22-    page: {
23:      title: `${doc.title} | Electron`,
24-      description: doc.description,

routes/docs/structures.js
12-    page: {
13:      title: 'API Structures | Electron'
14-    },

routes/docs/history.js
25-    page: {
26:      title: `${doc.title} Version History | Electron`,
27-      description: doc.description,

routes/feed/mainFeed.js
11-  let feed = new Feed({
12:    title: 'Electron',
13-    description:
--
40-          id: `https://electronjs.org${post.href}`,
41:          title: post.title,
42-          content: post.content,

routes/apps/show.js
16-    page: {
17:      title: `${app.name} | Apps | Electron`,
18-      url: req.url,
```

But that's okay, I also agree that it makes sense to localize this.

> What do you think about create mini blog post?

I like the idea! There are a couple other features I'd like to land first (after this PR lands) but let's definitely do it.

> It fixes #1071 and/or #1214?

I think it helps, but I think there's a little more work we can do here. In particular, it would still take a long time to find older maintenance releases under e.g. the stable releases. I'll experiment with a couple ideas here.

BinaryMuse commented:
@HashimotoYT I've pushed up an experimental change; what do you think?

People on specific (older) versions of Electron want to quickly find releases for that version; I added a filter to the bottom of the sidebar.

I've also changed the styling of the release type buttons a little; I found that the underline made it feel like we were splitting the element into two pieces, so I decided to try a left border. 💭?

![image](https://user-images.githubusercontent.com/189606/51281535-8c882700-1997-11e9-8991-f24a5a0eee83.png)

HashimotoYT commented:
I check it this morning. (time zones, yep)

BinaryMuse commented:
> I check it this morning. (time zones, yep)

Haha, no problem! We have a lot of time zones in this team, we're used to it ;)

HashimotoYT commented:
I' also push experimental change.

This change makes sidebar and main releases section into one component instead of using absolute position, it's fix some bugs with page viewing. Also i change container from narrow to xl, it gives user more information, and prevent moving sidebar to right what is so quite noticeable change.

How look xl container:

![how it look now](https://user-images.githubusercontent.com/24681191/51326580-190a1680-1a78-11e9-8f83-9c0370e5d6e5.png)

Bad styling it, should also contain versions from 3.x to 1.x:

![bad styling](https://user-images.githubusercontent.com/24681191/51326598-245d4200-1a78-11e9-9a6a-b304873214bf.png)

And maked TODO in releases styles.

BinaryMuse commented:
Alright, with the latest commit, entering a negative number or a number greater than the highest valid page number will redirect you to a page within the correct range. I think this is a decent behavior.

HashimotoYT commented:
@50Wliu I decide to refuse this idea, because I don't view at the moment need of making this sidebar sticky, and this idea came to me by chance.

BinaryMuse commented:
Thanks very much @HashimotoYT for all your work and help on this

zeke commented:
👏
